File: BayesianOptimization/spdn/spdnresult.py
from logger.csvwriter import CsvWriter

class SPDNResult:
    def __init__(self, value, point, algo_id, hyper_params, model):
        self.resultValue = value
        self.resultPoint = point
        self.algorithmId = algo_id
        self.hyperParams = hyper_params
        self.model = model
        self.csv_writer = CsvWriter(self.model.id, self.algorithmId)
        self.execution_time = 0

    def print_result(self):
        print("Algorithm: " + self.algorithmId)
        print("Model: " + self.model.id)
        print("Value = " + str(self.resultValue))
        for i in range(len(self.model.parameters)):
             print(self.model.parameters[i] + " = " + str(self.resultPoint[i]))

    def write_out_to_csv(self):
        row = ['MIN VALUE']
        for param in self.model.parameters: row.append(param + '(' + str(self.model.validvalues[param]) + ')')
        for hyperparam in self.hyperParams: row.append("HP_" + hyperparam)
        row.append("EXEC TIME")
        self.csv_writer.write(row, header=True)

        row = [str(self.resultValue)]
        for i in range(len(self.model.parameters)): row.append(str(self.resultPoint[i]))
        for hyperparam in self.hyperParams: row.append(str(self.hyperParams[hyperparam]))
        row.append(str(self.execution_time))
        self.csv_writer.write(row)

    def __del__(self): self.csv_writer.close()

File: NonConstrainedOptimization/src/main/java/spdn/SPDNResult.java
package spdn;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.List;
import java.util.SortedMap;

import hu.bme.mit.inf.petridotnet.spdn.Parameter;
import models.Model;
import spdn.model.SpdnModel;
import spdn.model.SpdnParameter;

public class SPDNResult {
	private double resultValue;
	private List<Double> resultPoint;
	private String algorithmID;
	private SortedMap<String,Double> hyperParams;
	private SpdnModel model;
	private PrintWriter csvWriter = null;
	private long time;
	
	private final boolean DONT_WRITE_TO_CSV = true;
	
	public SPDNResult(double value, List<Double> result, String aId, SortedMap<String,Double> hyperParams, SpdnModel model) {
		this.resultValue = value;
		this.resultPoint = result;
		this.algorithmID = aId;
		this.model = model;
		this.hyperParams = hyperParams;
	}
	
	public double getValue() { return resultValue; }
	
	public List<Double> getPoint() { return resultPoint; }
	
	public void setTime(long time) { this.time = time; }
	
	@Override
	public String toString() {
		String s = "Done in " + time / 60. / 1000000000. + " min\nAlgorithm: " + algorithmID + "\nModel: " + model.id + "\nValue = " + resultValue;
		if (resultPoint == null) resultPoint = new ArrayList<>(model.parameterSize());
		for (int i = 0; i < model.parameterSize(); i++) {
			s += "\n" + model.simpleParameterList.get(i).getName() + " = " + resultPoint.get(i);
		}
		return s;
	}
	
	public void writeToCsv() {
		if (DONT_WRITE_TO_CSV) return;
		try {
			String name = System.getProperty("user.dir");
			File f;
			if (name.contains("\\")) {
				f = new File(name + "\\results\\" +algorithmID + "-" + model.id + ".csv");
			} else {
				f = new File(name + "/results/" +algorithmID + "-" + model.id + ".csv");
			}
			boolean exist = f.exists();
			csvWriter = new PrintWriter(new FileOutputStream(f, true));
			
			if (!exist) {
				writeCsvHeader();
			}
			
			writeResultToCsv();
			
		} catch (IOException e) {
			System.err.println(e.getMessage());
		} finally{
			csvWriter.close();
		}
	}
	
	private void writeCsvHeader() {
		String header = "MIN VALUE;";
		for(SpdnParameter param: model.SpdnParameterList) {
			header += param.parameter.getName() + "(" + param.defaultValue + ");";
		}
		for(String key: hyperParams.keySet()) {
			header += "HP_" + key + ";";
		}
		header += "EXEC TIME(min)";
		csvWriter.append(header + "\n");
	}
	
	private void writeResultToCsv() {
		String row = String.valueOf(resultValue) + ";";
		if (resultPoint == null) resultPoint = new ArrayList<>(model.parameterSize());
		for(double x: resultPoint) {
			row += String.valueOf(x) + ";";
		}
		for(Double hyperparam: hyperParams.values()) {
			row += hyperparam + ";";
		}
		row += String.valueOf(time / 60. / 1000000000.);
		
		csvWriter.append(row + "\n");
	}
}
